streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableProcessorNode.java

```diff
@@ -64,8 +67,12 @@ public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
             topologyBuilder.connectProcessorAndStateStores(processorName, storeNames);
         }
 
-        // TODO: we are enforcing this as a keyvalue store, but it should go beyond any type of stores
-        if (storeBuilder != null) {
+        if (processorParameters.processorSupplier() instanceof KTableSource) {
+            if (((KTableSource<?, ?>) processorParameters.processorSupplier()).materialized()) {
+                topologyBuilder.addStateStore(Objects.requireNonNull(storeBuilder, "storeBuilder was null"),
+                                              processorName);
+            }
+        } else if (storeBuilder != null) {
```

Comment:
Ok, I figured out how to implement @mjsax 's suggestion specifically for KTableSource nodes, without affecting all the other kinds of KTable nodes.

Reply:
In retrospect, that might have been what you meant all along. Sorry if I misunderstood.

Reply:
As a further thought, I think TableProcessorNode should be used for KTableSource as well (today they are only used in filter, map, transformValues and joinForeignKey that results in a KTable), so that we do not need this extra condition. But we can do this cleanup later (our current internal representation has a few such gaps already so some refactoring might be in request in future anyways).

Reply:
I agree with you, and I actually tried to do it at some point. There's a problem with the types because (IIRC) TableProcessorNodes are processors over `Change<V>`, but TableSourceNodes are processors over `<V>`. I wanted to revisit it after KIP-478, but never managed to get around to actually implementing it (yet).